Inline runtime generation, fix test:browser

zapu opened the pull request:
Added inline runtime generation with `build:inline_runtime` task. The runtime is generated to `extras/inline-runtime.js` and is used in `nodes.coffee` to emit IcedRuntime, if the option is set to `inline` or `window`.

This method, however, limits compilation capabilities of iced as browser library. Iced in browser will not be able to emit inline runtime, because there will be no `inline-runtime.js` text file anywhere. Compiling with `runtime: none` and letting it use runtime from `iced` global variable will work - and that's what `test:browser` Cake task does right now.

All tests pass, for both `test` and `test:browser`. Keep in mind that `test:browser` runs less tests, grep `testingBrowser` to see which test files are skipped.

zapu committed:
Compile inline runtime to inline-runtime.js

Added Cakefile task to build inline runtime (iced-runtime-3) and save it
to extras/inline-runtime.js. It is used in nodes.coffee when runtime:
'inline' is requested.

zapu committed:
Fix inline runtime and browser tests

Improve IcedRuntime node to properly generate 'inline' and 'window'
runtimes. It is done by including extras/inline-runtime.js file into the
output. This file is generated with Cakefile build:inline_runtime task.

Fix browser tests. When in test:browser task, test files will be
compiled with runtime set to 'none'. `global.iced` will point to
iced-runtime instead. This is done to simulate browser environment.

zapu commented on Cakefile:
```diff
@@ -166,6 +169,30 @@ task 'build:browser', 'rebuild the merged script for inclusion in the browser',
+
+  unless process.env.MINIFY is 'false'
+    {code} = require('uglify-js').minify code, fromString: true
+
```
Can someone check if uglifying works? It fails for me in both build:browser and build:inline_runtime with parse errors.

maxtaco replied:
Yeah, no dice: https://github.com/mishoo/UglifyJS2/issues/448

zapu replied:
Aww, so it's related to es6 features? Ok...

maxtaco replied:
Yeah, seems to be puking on `function*`

maxtaco replied:
Talking to @chrisnojima about this in the office, we probably shouldn't be outputting ES6 to browsers anyways. People should convert to ES6 via babel (or whatnot).  This is of course a much deeper concern, and indicates a deeper change in workflow, so maybe we shouldn't spend too many cycles on this.

zapu replied:
Runtime can also be rewritten in pure coffeescript (manually doing the deferral calls etc).

maxtaco replied:
Hm that's a good suggestion.  Or just move `library.iced` into a separate repo, since it's not strictly required.

maxtaco replied:
However, if we did, it still wouldn't change the fact that the actual code emittted from iced3 is in ES6 (with `function*`) and will break  many browsers.

zapu replied:
But this is the point of this branch, no? To emit ES6 for future browsers or as input for babel translation.

We can optionally support browserify in compiler - but do not actually specify it as a dependency - so it will simplify the workflow for people who have it installed anyways.

maxtaco replied:
Agreed all around.  Probably the most useful workflow is for us to output ES6, and then for consumers of the output code to minify and transpile as they see fit. Meaning, the whole direct browser operation of CoffeeScript might be moot for us, or more charitably, best-effort.

maxtaco replied:
In sum, let's stick with the previous plan of (1) outputting the inline-runtime via raw string inclusion, and add (2) disabling of minification

zapu replied:
Cool, thanks! I have (1) almost done but I need to write some tests. I'll try to have it all committed here at the end of the day.

zapu commented on Cakefile:
```diff
@@ -166,6 +169,30 @@ task 'build:browser', 'rebuild the merged script for inclusion in the browser',
   console.log "built ... running browser tests:"
   invoke 'test:browser'
 
+task 'build:inline_runtime', 'build the inline iced3 runtime', ->
```
Should it be `build:inline_runtime` or `build:inline-runtime` ?

maxtaco replied:
the latter

zapu commented:
So what are we doing with this one?

maxtaco commented on src/nodes.coffee:
```diff
@@ -2437,6 +2438,10 @@ class IcedRuntime extends Block
     else               super o
 
+  inlineRuntime: (lefthand) ->
+    fs = require('fs')
```
i thought we can't use `require('fs')` on the browser?

maxtaco replied:
My only frustration is that we used to have something that did work here.  So how do you think the browser compiler should work then --- the user has to insert their own runtime?  I'm getting a little confused.

zapu commented:
About `require('fs')` in browser - I think you removed the comment while I was typing the reply. It might be useful anyways:

That's correct, this would be a limitation of the compiler in browser for now. It wouldn't be able to compile code with runtime set to 'inline'. I don't really see a good way around it. 

I tried to make the compiler "decompile" it's own runtime and emit it, but it led to nowhere. I can try this again if it sounds like a reasonable solution to you.

maxtaco commented:
Here's an idea I had:
- In the `cake build:inline-runtime` step, output a `lib/coffee-script/inline-runtime.js` file that has the code from `inline-runtime.js` in a raw string
- Emit the raw string on compilation in inline mode, instead of reading the file off of the fs via `require('fs')`. So instead, something of the form:

``` coffee
inlineRuntime: (lefthand) ->
    inline_runtime = require('./inline-runtime).runtime
    "#{lefthand} = #{inline_runtime};"
```

zapu commented:
OK, that was another idea that I've had before. I wasn't sure which one was better - decompiling or saving raw string. So I'll get to it later today.

zapu commented:
Just for reference (and for other people interested, maybe), `iced2` used to do this: https://github.com/maxtaco/coffee-script/blob/iced2/src/nodes.coffee#L3394 which was not ideal either.

maxtaco commented:
(I editted my comment above, thank you @zapu!)

maxtaco commented:
Indeed, the previous solution was bad for several reasons, especially because the code was different/worse and drifted further apart over time.

zapu committed:
call handleIcedOptions in command.coffee

zapu committed:
Save inline runtime in string form

Inline runtime will now contain stringified copy of itself. This is done
in order to be able to emit the runtime from a browser-based compiler
(or any environment that does not have a filesystem or access to
compiler files).

Removed uglifyjs support as it did not work with ES6 code.

Renamed "build:inline_runtime" task to "build:inline-runtime"

zapu commented:
I added the commit, it's still pretty rough around the edges, but the tests pass and seems OK overall.

zapu commented on src/nodes.coffee:
```diff
@@ -2437,6 +2439,26 @@ class IcedRuntime extends Block
+      # If not (compiler is stand alone), use readFileSync to get the
+      # runtime.
+      fs = require('fs')
+      runtime_str = fs.readFileSync "extras/inline-runtime.js", 'utf-8'
```
Oh, oops, this part is still not finished. And you have a solution for this too, but I misread your comment, sorry! I'll move the generation to `lib/` instead of `extras/` and fix this as well.

zapu commented:
Unsure if I haven't messed all of this up, working on this issue feels like programming a quine :confounded:

zapu committed:
Save inline runtime in lib/ instead of extras/

Stringified inline runtime is now accessible via
require('./inline-runtime-str') from both node version and the browser.
Much cleaner solution than the previous one.

zapu commented:
Changed this approach a little bit to be more lined up with your proposal, and removed irrelevant comments from this PR. Things are much cleaner right now.

I should also squash the commits before eventual merge.

maxtaco commented:
Looking very good. Do you get all tests passing?  I'm still seeing one failing (repl.coffee:69).

maxtaco commented:
```
 node ./bin/cake test

failed 1 and passed 714 tests in 2.58 seconds 

  empty command evaluates to undefined 
  AssertionError: Expected undefined to equal iced> 
  at global.eq (/Users/max/src/iced/coffee-script/Cakefile:363:14)
  at /Users/max/src/iced/coffee-script/test/repl.coffee:69:3
  at Function.<anonymous> (/Users/max/src/iced/coffee-script/test/repl.coffee:38:17)
  at global.test (/Users/max/src/iced/coffee-script/Cakefile:287:12)
  at testRepl (/Users/max/src/iced/coffee-script/test/repl.coffee:38:3)
  at Object.<anonymous> (/Users/max/src/iced/coffee-script/test/repl.coffee:67:1)
  at Object.<anonymous> (/Users/max/src/iced/coffee-script/test/repl.coffee:1:1)
  at Module._compile (module.js:409:26)
  at Object.exports.run (/Users/max/src/iced/coffee-script/lib/coffee-script/coffee-script.js:136:23)
  at runTests (/Users/max/src/iced/coffee-script/Cakefile:412:22)
  at Object.action (/Users/max/src/iced/coffee-script/Cakefile:429:12)
  at helpers.extend.invoke (/Users/max/src/iced/coffee-script/lib/coffee-script/cake.js:44:26)
  at Object.exports.run (/Users/max/src/iced/coffee-script/lib/coffee-script/cake.js:70:20)
  at Object.<anonymous> (/Users/max/src/iced/coffee-script/bin/cake:7:38)
  at Module._compile (module.js:409:26)
  at Object.Module._extensions..js (module.js:416:10)
  at Module.load (module.js:343:32)
  at Function.Module._load (module.js:300:12)
  at Function.Module.runMain (module.js:441:10)
  at startup (node.js:134:18)
  at node.js:962:3

  function () {
      return fn(input, output, repl);
    }
```

zapu commented:
Hm, there was a change in behavior in node but I'm no longer seeing that. I tried to make this change in upstream (https://github.com/jashkenas/coffeescript/pull/4120) but it wasn't accepted. What node version do you have? I'll try to find some time and dig through the changelogs.

Can you try something like this?

```
zapu@lab » node
> undefined
undefined
> console.log('hell')
hell
undefined
>
zapu@lab » node --version
v5.6.0
```

Also, what's with the versions? I guess their versioning changed after all the node.js/io.js things happened.

maxtaco commented:
Ah, I'm running `4.3.2`....

zapu commented:
The correct way would be to detect repl settings (because I think you can change the undefined behavior) and make test check result according to this setting.

maxtaco commented:
Well still broken for me on 5.10.1., but that's ok.  

:+1: Ok, to merge!

zapu commented:
Oh, cool! Thanks!

maxtaco commented:
Thank you!
